Добавление файлового менеджера

Sayders11 opened the pull request:
1) Добавил новую реализацию менеджера - файловую

2) Реализовал сохранение данных в файл формата CSV

3) Добавил возможность загрузки данных из файла в менеджер

Makhoro committed:
1) Добавил новую реализацию менеджера - файловую

2) Реализовал сохранение данных в файл формата CSV

3) Добавил возможность загрузки данных из файла в менеджер

Makhoro committed:
1) Добавил новую реализацию менеджера - файловую

2) Реализовал сохранение данных в файл формата CSV

3) Добавил возможность загрузки данных из файла в менеджер

Makhoro committed:
1) Добавил новую реализацию менеджера - файловую

2) Реализовал сохранение данных в файл формата CSV

3) Добавил возможность загрузки данных из файла в менеджер

Makhoro committed:
1) Добавил новую реализацию менеджера - файловую

2) Реализовал сохранение данных в файл формата CSV

3) Добавил возможность загрузки данных из файла в менеджер

4) Покрыл новый менеджер тестами

Makhoro committed:
1) Добавил новую реализацию менеджера - файловую

2) Реализовал сохранение данных в файл формата CSV

3) Добавил возможность загрузки данных из файла в менеджер

Makhoro committed:
1) Добавил новую реализацию менеджера - файловую

2) Реализовал сохранение данных в файл формата CSV

3) Добавил возможность загрузки данных из файла в менеджер

LexLippi commented on src/model/Subtask.java:
```diff
@@ -18,6 +22,10 @@ public void setEpicId(int epicId) {
         this.epicId = epicId;
     }
 
+    public Type getType() {
+        return type;
+    }
```
Нет необходимости в этом методе, тем более не переопределенном, он есть в классе родителе

Sayders11 replied:
done

LexLippi commented on src/model/Epic.java:
```diff
@@ -14,6 +15,10 @@ public ArrayList<Integer> getSubtasksIds() {
         return subtasksIds;
     }
 
+    public Type getType() {
+        return type;
+    }
```
Нет необходимости в этом методе, тем более не переопределенном, он есть в классе родителе

Sayders11 replied:
done

LexLippi commented on src/service/FileBackedTaskManager.java:
```diff
@@ -0,0 +1,200 @@
+                        break;
+                }
+
+                if (fileManager.seq < task.getId()) {
+                    fileManager.seq = task.getId();
+                }
```
Можно использовать метод `max` из класса `java.lang.Math`

Sayders11 replied:
done

LexLippi commented on src/service/FileBackedTaskManager.java:
```diff
@@ -0,0 +1,200 @@
+
+    public static FileBackedTaskManager loadFromFile(File file) {
+
+        FileBackedTaskManager fileManager = new FileBackedTaskManager(new InMemoryHistoryManager());
```
лучше создавать с переданным файлом, который мы в аргументе метода принимаем, иначе писать мы всегда будем в один файл, что не очень хорошо

LexLippi commented on src/service/FileBackedTaskManager.java:
```diff
@@ -0,0 +1,200 @@
+                }
+            }
+        } catch (IOException e) {
+            throw new RuntimeException(e);
```
Лучше тоже кастомное исключение бросить

LexLippi commented on src/service/FileBackedTaskManager.java:
```diff
@@ -0,0 +1,200 @@
+        }
+    }
+
+    private String toString(Task task) {
+        return task.getId() + "," + task.getType() + "," + task.getName() + "," + task.getStatus() + ","
+                + task.getDescription();
+    }
+
+    private String toString(Epic epic) {
+        return epic.getId() + "," + epic.getType() + "," + epic.getName() + "," + epic.getStatus() + ","
+                + epic.getDescription();
+    }
+
+    private String toString(Subtask subtask) {
+        return subtask.getId() + "," + subtask.getType() + "," + subtask.getName() + "," + subtask.getStatus() + ","
+                + subtask.getDescription() + "," + subtask.getEpicId();
+    }
```
Здесь лучше использовать полиморфизм и перенести эти методы в соответствующие классы, добавив аннотацию `Override`. Примерно так:

```java
public String toString() {
    return this.getId() + "," + this.getType() + "," + this.getName() + "," + this.getStatus() + ","
                + this.getDescription();
}
```

LexLippi commented on src/service/FileBackedTaskManager.java:
```diff
@@ -0,0 +1,200 @@
+        String taskName = parts[2];
+        String taskStatus = parts[3];
+        String taskDescription = parts[4];
+        Task returnedTask = null;
```
Я противник инициализации переменных при помощи `null`, когда это возможно.
Здесь можно в каждой ветке `case` делать `return backTask;` вместо `break`

LexLippi reviewed (changes requested):
Привет!

Хорошая работа!

Есть некоторое количество замечаний, которые необходимо исправить!

Желаю удачи!

Makhoro committed:
Внес правки в код:

1) Убрал лишние методы
2) Заменил условный оператор на метод Math
3) Улучшил метод loadFromFile
4) Заменил стандартные исключения на кастомные
5) Перенес методы toString в другие классы согласно принципу полиморфизма
6) Поправил метод fromString, убрал из него инициализацию переменной

LexLippi commented:
Привет!

Отличная работа!

Желаю удачи в следующих спринтах!
